src/Generator/Passes/DelegatesPass.cs

```diff
@@ -96,36 +112,63 @@ public override bool VisitMethodDecl(Method method)
                             new QualifiedType(
                                 new FunctionType
                                 {
-                                    CallingConvention = method.CallingConvention,
-                                    IsDependent = method.IsDependent,
+                                    CallingConvention = callingConvention,
+                                    IsDependent = isDependent,
                                     Parameters = @params,
-                                    ReturnType = method.ReturnType
+                                    ReturnType = returnType
                                 }))),
                     Namespace = namespaceDelegates,
                     IsSynthetized = true,
-                    Access = AccessSpecifier.Private
+                    Access = (decl is Method) ? AccessSpecifier.Private : AccessSpecifier.Public
                 };
 
             var delegateString = @delegate.Visit(TypePrinter).Type;
             var existingDelegate = GetExistingDelegate(
-                method.TranslationUnit.Module, delegateString);
+                module, delegateString);
 
             if (existingDelegate != null)
             {
-                Context.Delegates.Add(method, existingDelegate);
-                return true;
+                Context.Delegates.Add(decl, existingDelegate);
+                return;
             }
 
             existingDelegate = new DelegateDefinition(module.OutputNamespace, delegateString);
-            Context.Delegates.Add(method, existingDelegate);
+            Context.Delegates.Add(decl, existingDelegate);
 
             libsDelegates[module].Add(delegateString, existingDelegate);
 
             namespaceDelegates.Declarations.Add(@delegate);
+        }
 
-            return true;
+        private void VisitFunctionTypeParameters(Method method)
+        {
+            foreach (var param in method.Parameters.Where(
```

Comment:
Let me just tell you how I want it written:

foreach (var param in from param in method.Parameters
                                     let paramType = param.Type.Desugar()
                                     where paramType.IsAddress() &&
                                         paramType.GetFinalPointee().Type.Desugar() is FunctionType)